infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,135 @@
+import static org.eclipse.che.workspace.infrastructure.kubernetes.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.time.Clock;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.user.server.PreferenceManager;
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.workspace.server.WorkspaceRuntimes;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalRuntime;
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class AsyncStoragePodWatcher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodWatcher.class);
+
+  private final KubernetesClientFactory kubernetesClientFactory;
+  private final UserManager userManager;
+  private final PreferenceManager preferenceManager;
+  private final WorkspaceRuntimes runtimes;
+  private final long shutdownTimeoutSec;
+  private final Clock clock = Clock.systemDefaultZone();
+
+  @Inject
+  public AsyncStoragePodWatcher(
+      KubernetesClientFactory kubernetesClientFactory,
+      UserManager userManager,
+      PreferenceManager preferenceManager,
+      WorkspaceRuntimes runtimes,
+      @Named("che.infra.kubernetes.async.storage.shutdown_timeout_min") long shutdownTimeoutMin) {
+    this.kubernetesClientFactory = kubernetesClientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+    this.runtimes = runtimes;
+    this.shutdownTimeoutSec = shutdownTimeoutMin * 60;
+  }
+
+  @ScheduleDelay(
+      unit = TimeUnit.MINUTES,
+      initialDelay = 1,
+      delayParameterName = "che.infra.kubernetes.async.storage.shutdown_check_period_min")
+  public void check() {
+    for (User user :
+        iterateLazily((maxItems, skipCount) -> userManager.getAll(maxItems, skipCount))) {
+      try {
+        String owner = user.getId();
+        Map<String, String> preferences = preferenceManager.find(owner);
+        String lastTimeAccess = preferences.get(LAST_ACTIVITY_TIME);
+        String namespace = preferences.get(LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE);
+        if (isNullOrEmpty(namespace)
+            || isNullOrEmpty(lastTimeAccess)
+            || isAnyRuntimeInProgress(owner)) {
+          continue;
+        }
+        long lastTimeAccessSec = parseLong(lastTimeAccess);
+        long epochSec = clock.instant().getEpochSecond();
```

Comment:
I would say shutdownTimeoutSec-> shutdownTimeoutMsec 
and instead of clock.instant().getEpochSecond() use System.currentTimeMillis()